sdk/python/lib/pulumi/dynamic/dynamic.py

```diff
@@ -247,6 +250,7 @@ def __init__(self,
         if PROVIDER_KEY in props:
             raise  Exception("A dynamic resource must not define the __provider key")
 
+        props = cast(dict, props)
```

Comment:
so this means: from this point on, think of `props` as a `dictionary`?

Reply:
It's only for typechecking purposes. While this runs at runtime, it is effectively a noop. If you look at the surrounding code, there's some confusion in the docstring vs the type signature of props. One says `Optional[dict]` and the other `Inputs`. Mappings in python don't seem to be indexable, which is why we need to do the cast to enable `props[PROVIDER_KEY] on the next line.